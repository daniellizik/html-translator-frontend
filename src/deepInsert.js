/**
 * deepInsert - walks through list and mutates each corresponding node in origin tree
 * 
 * 
 * 
 * 
 * 
 * 
 */

export default function deepInsert() {

  return function walk() {

  }

}