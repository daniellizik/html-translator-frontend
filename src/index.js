import React from 'react'
import { render } from 'react-dom'
import { Provider } from 'react-redux'
import { StyleRoot } from 'radium'
import Layout from './containers/layout'
import { configureStore } from './store/configureStore'
const { store } = configureStore()

render(
  <StyleRoot>
    <Provider store={store}>
      <Layout />
    </Provider>
  </StyleRoot>,
  document.getElementById('root')
)
