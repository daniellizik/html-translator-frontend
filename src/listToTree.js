const defaultConfig = {
  childrenProp: 'childNodes',
  attributesProp: 'attrs',
  nodeNameProp: 'nodeName',
  valueProp: 'value',
  parentProp: 'parent',
  idProp: 'id'
}

export default function treeToList(config = defaultConfig) {
  return function walk(list = [], curr = 0, tree = {...list[0], [config.childrenProp]: []}) {
    return {
      ...tree,
      [config.childrenProp]: [
        ...tree[config.childrenProp],
        // bad time complexity
        list.find(c => c[config.parentProp] === c[config.idProp])
      ]
    }
  }
}