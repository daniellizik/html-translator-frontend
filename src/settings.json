{
  "onboarding_fade_delay": 700,
  "onboarding_fade_max": 3
}