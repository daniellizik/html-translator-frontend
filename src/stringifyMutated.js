export default function(xml) {
  return xml.reduce((acc, {node, row, tokens}) => {
    const depth = new Array(node.depth).fill('  ').join('')
    const line = tokens.map(t => t.value).join('')
    return [...acc, depth + line]
  }, []).join('\n')
}