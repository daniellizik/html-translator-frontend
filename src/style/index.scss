@import './_colors.scss';
@import './_typography.scss';
@import './_helpers.scss';
@import './_override.scss';
@import './rc-tooltip.scss';
@import './_code.scss';