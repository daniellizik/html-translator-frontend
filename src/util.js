export const chainReducers = (state, action, ...reducers) => {
  return reducers.reduce((reducedState, reducer) => {
    const result = reducer(reducedState, action)
    return result
  }, state)
}

export const chainActions = (initialState, reducer, ...actions) => {
  const state = Array.isArray(initialState) ? initialState.slice(-1).pop() : initialState
  return actions.reduce((acc, action) => {
    const result = reducer(acc.state, action)
    return {
      state: result,
      history: [...acc.history, result]
    }
  }, {state, history: [state]}).history
}

export const asyncChainActions = (store, ...actions) => {
  return actions.reduce((acc, action) => {

  }, [])
}

export const filterText = (node) => {
  return !/^[\s\r\n]+$/.test(node.value)
}

export const mutateList = (list, view) => {
  return list.reduce((acc, node) => { 
    const index = view.indexOf(node.id)
    return index < 0 
      ? [...acc, node]
      : [...acc, {...node, view: true}]
  },[])
}

export const bindConstantsToReducers = (reducers, initialState) => (state, action) => {
  return reducers[action.type] 
    ? reducers[action.type](state, action) 
    : (state || initialState)
}