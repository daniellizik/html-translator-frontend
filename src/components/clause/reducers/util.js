import * as constants from '../constants'
import * as targets from '../settings/targets'
import * as rules from '../settings/rules'
import * as behaviors from '../settings/behaviors'
import { targetMap } from '../settings/config'

export const errorHandler = ({message}) => {
  return message
}

export const mutateList = (list = [], view = []) => {
  return list.reduce((acc, node) => { 
    const index = view.indexOf(node.id)
    return index < 0 
      ? [...acc, node]
      : [...acc, {...node, view: true}]
  },[])
}

// take view, mutations and apply them to FULL LIST, not open
// this makes it easier for xml tree to consume (it just spits out entire list)
export const mutationDenormalizer = (clause, list = []) => {
  const { target, view, mutations } = clause
  // mutations might have to mutate entire list, so give a copy to keep it immutable
  return list.slice().reduce((acc, node, i, arr) => {
    // only mutate items in view
    return view.indexOf(node.id) < 0 ? [...acc, node] : [
      ...acc,
      // this needs to account for mutation behavior
      mutations.reduce((mutatedNode, mutation) => {
        const targetProp = targetMap[target]
        const params = {...mutation, before: mutatedNode[targetProp]}
        const ruleResult = rules[mutation.rule](params)
        // todo: implement this
        // const behaviorResult = behaviors[mutation.behavior](mutatedNode, mutation, arr, ruleResult)
        return {
          ...mutatedNode,
          [targetProp]: ruleResult
        }
      }, {...node})
    ] 
  }, [])
}

export const reduceRuleProp = (ruleType, state, action, prop) => {
  return {
    ...state,
    clauses: state.clauses.map((c, i) => (
      i !== action.clauseIndex ? c : {
        ...c,
        [ruleType]: c[ruleType].map((r, j) => (
          j !== action.ruleIndex ? r : {
            ...r,
            [prop]: action[prop]
          }
        ))
      }
    ))
  }
}

export const reduceView = (clauseIndex, {list}, clauses) => {
  return clauses.reduce((acc, clause, index) => {
    if (index !== clauseIndex)
      return [...acc, clause]
    const view = list.open.reduce((ids, node) => {
      const clauseResult = clause.queries.reduce((bool, obj) => {
        const targetResult = targets.query[clause.target](node, {...obj, rule: rules[obj.rule]})
        if (targetResult === false || bool === false)
          return false
        else if (targetResult === true)
          return true
      }, null)
      return !clauseResult ? ids : [...ids, node.id]
    }, [])
    return [
      ...acc,
      { ...clause, view }
    ]
  }, [])
}

export const reduceClauses = (state, action, key) => {
  let nextState
  const clauses = state.clauses.map((clause, clauseIndex) => {
    return clauseIndex !== action.clauseIndex ? clause : {
      ...clause,
      queries: clause.queries.map((query, ruleIndex) => {
        return ruleIndex !== action.ruleIndex ? query : { ...query, [key]: action[key] }
      })
    }
  })
  try {
    nextState = {
      ...state,
      clauses: reduceView(action.clauseIndex, state.slave, clauses)
    }
  }
  catch(e) {
    process.env.NODE_ENV === 'development' && console.warn('error', e)
    return {
      ...state,
      clauses,
      error: errorHandler(e)
    }
  }
  return nextState
}