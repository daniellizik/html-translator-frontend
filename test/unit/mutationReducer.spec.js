import { omit } from '~/src/util'
import * as mutations from '~/test/storyFixtures/mutation'

const fixtures = Object.keys(mutations).reduce((acc, m) => {
  return {
    ...acc,
    [m]: mutations[m].map(s => {
      return {
        ...s,
        slave: omit(s.source, 'rawHtml')
      }
    })
  }
}, {})

describe('mutation reducer', () => {

  describe('replace all text', () => {
    it('should replace all text', () => {
      expect(fixtures.allReplaceText.pop()).toMatchSnapshot()
    })
  })

  describe('reduce mutations', () => {
    it('should reduce mutations', () => {
      expect(fixtures.reducedMutations.pop()).toMatchSnapshot()
    })
  })

  describe('regex mutation', () => {
    it('should mutate with regex', () => {
      expect(fixtures.regexMutation.pop()).toMatchSnapshot()
    })
  })

  describe('toggling', () => {
    it('should toggle view', () => {
      expect(fixtures.toggling.pop()).toMatchSnapshot()
    })
  })

  describe('mutating nodename', () => {
    it('should mutate nodename', () => {
      expect(fixtures.nodename.pop()).toMatchSnapshot()
    })
  })

 describe('chain denormalizations', () => {
    it('should chain', () => {
      expect(fixtures.chainedDenormalizations.pop()).toMatchSnapshot()
    })
  })

})
