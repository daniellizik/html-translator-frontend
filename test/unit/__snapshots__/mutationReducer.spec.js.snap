exports[`mutation reducer chain denormalizations should chain 1`] = `
Object {
  "activeClause": -1,
  "clauses": Array [
    Object {
      "active": true,
      "minimized": false,
      "mutations": Array [
        Object {
          "active": true,
          "behavior": "",
          "rule": "ALL_REPLACE",
          "ruleValue": "blah",
          "ruleValueFlags": "",
          "targetValue": "",
        },
        Object {
          "behavior": "",
          "rule": "END_OF",
          "ruleValue": "-foo",
          "ruleValueFlags": "",
          "targetValue": "",
        },
      ],
      "name": "blah",
      "queries": Array [
        Object {
          "rule": "LIKE",
          "ruleValue": "",
          "ruleValueFlags": "",
          "targetValue": "cat-",
          "type": "QUERY",
        },
      ],
      "target": "TEXT",
      "view": Array [
        12,
        15,
        18,
      ],
    },
    Object {
      "active": false,
      "minimized": false,
      "mutations": Array [
        Object {
          "behavior": "",
          "rule": "ALL_REPLACE",
          "ruleValue": "span-good",
          "ruleValueFlags": "",
          "targetValue": "",
        },
        Object {
          "behavior": "",
          "rule": "START_OF",
          "ruleValue": "",
          "ruleValueFlags": "",
          "targetValue": "",
        },
      ],
      "name": "",
      "queries": Array [
        Object {
          "rule": "EQUALS",
          "ruleValue": "",
          "ruleValueFlags": "",
          "targetValue": "span",
        },
      ],
      "target": "NODE_NAME",
      "type": "CLAUSE",
      "view": Array [],
    },
  ],
  "error": "@CLAUSE/CLAUSE_ERROR_INVALID_REGEX",
  "iframe": Object {
    "src": null,
    "state": false,
  },
  "locale": "en-US",
  "onboarding": Object {
    "step": -1,
  },
  "overlay": false,
  "pastInit": true,
  "slave": Object {
    "active": false,
    "lastModified": null,
    "name": undefined,
    "url": null,
  },
  "source": Object {
    "active": false,
    "lastModified": null,
    "name": undefined,
    "rawHtml": null,
    "url": null,
  },
  "user": Object {
    "auth": true,
    "maxClauses": Infinity,
    "viewingSettings": false,
  },
}
`;

exports[`mutation reducer mutating nodename should mutate nodename 1`] = `
Object {
  "activeClause": 0,
  "clauses": Array [
    Object {
      "active": true,
      "minimized": false,
      "mutations": Array [
        Object {
          "active": true,
          "behavior": "",
          "rule": "ALL_REPLACE",
          "ruleValue": "h13",
          "ruleValueFlags": "",
          "targetValue": "",
        },
      ],
      "name": "blah",
      "queries": Array [
        Object {
          "rule": "LIKE",
          "ruleValue": "",
          "ruleValueFlags": "",
          "targetValue": "span",
          "type": "QUERY",
        },
      ],
      "target": "NODE_NAME",
      "view": Array [
        22,
        25,
        28,
        31,
        34,
      ],
    },
  ],
  "error": "@CLAUSE/CLAUSE_ERROR_INVALID_REGEX",
  "iframe": Object {
    "src": null,
    "state": false,
  },
  "locale": "en-US",
  "onboarding": Object {
    "step": -1,
  },
  "overlay": false,
  "pastInit": true,
  "slave": Object {
    "active": false,
    "lastModified": null,
    "name": undefined,
    "url": null,
  },
  "source": Object {
    "active": false,
    "lastModified": null,
    "name": undefined,
    "rawHtml": null,
    "url": null,
  },
  "user": Object {
    "auth": true,
    "maxClauses": Infinity,
    "viewingSettings": false,
  },
}
`;

exports[`mutation reducer reduce mutations should reduce mutations 1`] = `
Object {
  "activeClause": 0,
  "clauses": Array [
    Object {
      "active": true,
      "minimized": false,
      "mutations": Array [
        Object {
          "active": true,
          "behavior": "",
          "rule": "ALL_REPLACE",
          "ruleValue": "blah",
          "ruleValueFlags": "",
          "targetValue": "",
        },
        Object {
          "behavior": "",
          "rule": "START_OF",
          "ruleValue": "cat ",
          "ruleValueFlags": "",
          "targetValue": "",
        },
      ],
      "name": "blah",
      "queries": Array [
        Object {
          "rule": "LIKE",
          "ruleValue": "",
          "ruleValueFlags": "",
          "targetValue": "cat",
          "type": "QUERY",
        },
      ],
      "target": "TEXT",
      "view": Array [
        12,
        15,
        18,
        23,
        26,
        29,
        32,
        35,
        40,
        43,
        46,
        49,
        52,
        55,
      ],
    },
  ],
  "error": null,
  "iframe": Object {
    "src": null,
    "state": false,
  },
  "locale": "en-US",
  "onboarding": Object {
    "step": -1,
  },
  "overlay": false,
  "pastInit": true,
  "slave": Object {
    "active": false,
    "lastModified": null,
    "name": undefined,
    "url": null,
  },
  "source": Object {
    "active": false,
    "lastModified": null,
    "name": undefined,
    "rawHtml": null,
    "url": null,
  },
  "user": Object {
    "auth": true,
    "maxClauses": Infinity,
    "viewingSettings": false,
  },
}
`;

exports[`mutation reducer regex mutation should mutate with regex 1`] = `
Object {
  "activeClause": 0,
  "clauses": Array [
    Object {
      "active": true,
      "minimized": false,
      "mutations": Array [
        Object {
          "active": true,
          "behavior": "",
          "rule": "REGEX_REPLACE",
          "ruleValue": "cat",
          "ruleValueFlags": "[mutate!!]",
          "targetValue": "peanuts",
        },
        Object {
          "active": true,
          "behavior": "",
          "rule": "REGEX_REPLACE",
          "ruleValue": "eanu",
          "ruleValueFlags": "",
          "targetValue": " foobar ",
        },
      ],
      "name": "blah",
      "queries": Array [
        Object {
          "rule": "LIKE",
          "ruleValue": "",
          "ruleValueFlags": "",
          "targetValue": "cat",
          "type": "QUERY",
        },
      ],
      "target": "TEXT",
      "view": Array [
        12,
        15,
        18,
        23,
        26,
        29,
        32,
        35,
        40,
        43,
        46,
        49,
        52,
        55,
      ],
    },
  ],
  "error": "@CLAUSE/CLAUSE_ERROR_INVALID_REGEX",
  "iframe": Object {
    "src": null,
    "state": false,
  },
  "locale": "en-US",
  "onboarding": Object {
    "step": -1,
  },
  "overlay": false,
  "pastInit": true,
  "slave": Object {
    "active": false,
    "lastModified": null,
    "name": undefined,
    "url": null,
  },
  "source": Object {
    "active": false,
    "lastModified": null,
    "name": undefined,
    "rawHtml": null,
    "url": null,
  },
  "user": Object {
    "auth": true,
    "maxClauses": Infinity,
    "viewingSettings": false,
  },
}
`;

exports[`mutation reducer replace all text should replace all text 1`] = `
Object {
  "activeClause": 0,
  "clauses": Array [
    Object {
      "active": true,
      "minimized": false,
      "mutations": Array [
        Object {
          "active": true,
          "behavior": "",
          "rule": "ALL_REPLACE",
          "ruleValue": "blah",
          "ruleValueFlags": "",
          "targetValue": "",
        },
      ],
      "name": "blah",
      "queries": Array [
        Object {
          "rule": "LIKE",
          "ruleValue": "",
          "ruleValueFlags": "",
          "targetValue": "cat",
          "type": "QUERY",
        },
      ],
      "target": "TEXT",
      "view": Array [
        12,
        15,
        18,
        23,
        26,
        29,
        32,
        35,
        40,
        43,
        46,
        49,
        52,
        55,
      ],
    },
  ],
  "error": null,
  "iframe": Object {
    "src": null,
    "state": false,
  },
  "locale": "en-US",
  "onboarding": Object {
    "step": -1,
  },
  "overlay": false,
  "pastInit": true,
  "slave": Object {
    "active": false,
    "lastModified": null,
    "name": undefined,
    "url": null,
  },
  "source": Object {
    "active": false,
    "lastModified": null,
    "name": undefined,
    "rawHtml": null,
    "url": null,
  },
  "user": Object {
    "auth": true,
    "maxClauses": Infinity,
    "viewingSettings": false,
  },
}
`;

exports[`mutation reducer toggling should toggle view 1`] = `
Object {
  "activeClause": 0,
  "clauses": Array [
    Object {
      "active": true,
      "minimized": false,
      "mutations": Array [
        Object {
          "active": true,
          "behavior": "",
          "rule": "REGEX_REPLACE",
          "ruleValue": "cat",
          "ruleValueFlags": "[mutate!!]",
          "targetValue": "peanuts",
        },
        Object {
          "active": true,
          "behavior": "",
          "rule": "REGEX_REPLACE",
          "ruleValue": "eanu",
          "ruleValueFlags": "",
          "targetValue": " foobar ",
        },
      ],
      "name": "blah",
      "queries": Array [
        Object {
          "rule": "LIKE",
          "ruleValue": "",
          "ruleValueFlags": "",
          "targetValue": "cat",
          "type": "QUERY",
        },
      ],
      "target": "TEXT",
      "view": Array [
        12,
        15,
        18,
        23,
        26,
        29,
        32,
        35,
        40,
        43,
        46,
        49,
        52,
        55,
      ],
    },
  ],
  "error": "@CLAUSE/CLAUSE_ERROR_INVALID_REGEX",
  "iframe": Object {
    "src": null,
    "state": false,
  },
  "locale": "en-US",
  "onboarding": Object {
    "step": -1,
  },
  "overlay": false,
  "pastInit": true,
  "slave": Object {
    "active": false,
    "lastModified": null,
    "name": undefined,
    "url": null,
  },
  "source": Object {
    "active": false,
    "lastModified": null,
    "name": undefined,
    "rawHtml": null,
    "url": null,
  },
  "user": Object {
    "auth": true,
    "maxClauses": Infinity,
    "viewingSettings": false,
  },
}
`;
